### PR TITLE
Updates to institution citation keys

### LINKED ISSUE #6942: Regex of University Institution too broad for citekey generation
```
JabRef 5.2--2020-09-22--129c36e
Windows 10 10.0 amd64 
Java 15
```

- [x]  **Mandatory**: I have tested the latest development version from http://builds.jabref.org/master/ and the problem persists

- **Steps to reproduce the behavior:**
    - Add an entry where the author is an institution (other than an university) and contains a term beginning with `uni` (see below for example)
    - Enclose the institution in {}
    - Let Jabref generate the `citekey` as an institution with default settings

- **Expected Behavior**
Generate the `citekey` as a regular institution abbrevation

- **Observed Behavior**
`citekey` is generated as a "University" citekey beginning with `Uni` and adding a short form of the abbrevation afterwards

- **Alternative Behavior**
Make use of the `shortauthor` field if present for that entry. `biblatex-apa` for example uses this field as the institution abbrevation, so if wished, one can add the shortauthor field to the respective entrytypes.

**Longer description:**
I tried to add the following two entries and let Jabref generate the `citekey` automatically. 

```
% Encoding: UTF-8

@Report{ICAO2013,
  author      = {{International Civil Aviation Organization}},
  date        = {2013},
  institution = {{International Civil Aviation Organization}},
  location    = {Montréal, Quebec},
  publisher   = {International Civil Aviation Organization},
  shortauthor = {ICAO},
  title       = {Foo},
  type        = {resreport},
}

@Report{UniEuropeanAviationSafetyAgency2019,
  author       = {{European Union Aviation Safety Agency}},
  date         = {2019-12-18},
  institution  = {{European Union Aviation Safety Agency}},
  title        = {Bar},
  type         = {resreport},
  shortauthor  = {EASA},
  organization = {{European Union Aviation Safety Agency}},
}
@Comment{jabref-meta: databaseType:biblatex;}
```

I did not understand why the automatically generation would work for the ICAO entry but would not work for the EASA field (I would have been okay with having EUASA as the `citekey`, as I was aware that the automatic abbrevation would have used the U in the initials of the name). Insted, a completely different key was generated.
I tried to pinpoint the issue and stumbled upon this Regex in the key generation for institutions in brackets.
https://github.com/JabRef/jabref/blob/129c36e24bbd3332832728124ebf27d2462a0300/src/main/java/org/jabref/logic/citationkeypattern/BracketedPattern.java#L70-L79
To me, the Regex seems a bit broad but maybe this was on purpose. If so, I would be happy if there was an option to somehow have a setting to use the optional shortauthor field if present.
Unfortunately, I'm not skilled enough to implement a fix in a PR but I wanted to point out that this might be an issue.

If someone asks: I do not use the suggested institution abbrevation mentioned [here](https://docs.jabref.org/advanced/fields) as this messes with the institution abbrevation in the `biblatex-apa` package which uses the shortauthor. And even with an added abbrevation behind the full name, the citekey is still wrong.

### LINKED ISSUE #7199: Chinese author of essay reference section displays extra space
Hi, i am using JabRef (JabRef 5.2--2020-12-14--e9be339 Linux 5.9.0-4-amd64 amd64 Java 15.0.1) on Debian sid.

JabRef version  on 

- [ ] **Mandatory**: I have tested the latest development version from http://builds.jabref.org/master/ and the problem persists

After importing the Chinese essay, manually insert a space or comma, such as 杨 秀群, the citationkey will generate successfully, but the Reference section of essay will display "杨 秀群 "(there is a space between杨 and 秀群) , not "杨秀群".
How could you resolve this problem, thanks!

image
The JabRef screenshot.
image
The corresponding pdf file screenshot.
And the bibliography file is 
image

### PR DESCRIPTION
**Pull request by k3KAW8Pnf7mkmdSMPHz27**
Fixes #6942. Fixes #7199.

#### TL;DR
Authors only having a last name are abbreviated. #6942 is miss-abbreviated because name parts containing `uni` are assumed to be universities, e.g., `United Airlines`. #7199 is triggers the abbreviation method because, from BibTeX´s point-of-view, the author only has a last name. The name is later removed because Java does not classify the first character as an uppercase letter, hence the heuristic assumes it is an insignificant word. E.g., `{eBay}` gets removed, `{JabRef}` doesn't.

#### Background
Authors that are institutions can have very lengthy citation keys unless they are abbreviated in some fashion. The abbreviations are generated in

https://github.com/JabRef/jabref/blob/6718c08cf5ace13365da2ce4c1509e4817eb34bc/src/main/java/org/jabref/logic/citationkeypattern/BracketedPattern.java#L1243

which is based on a heuristic that tries to determine a suitable abbreviation based on the content of the specific author.

Issue #6942 is due to a regexp being overly broad when determining if an author is a university.

Issue #7199 was created from #6706 where I changed how a company/institute is identified. In previous versions of JabRef, an author is an institute if it is enclosed in curly brackets (e.g., `{JabRef}`), and I changed it to also include authors' who only have a last name (e.g., `JabRef`), which means that the author in issue #7199 is treated as an institution.
This would not have been an issue if it wasn't because the heuristic for creating a citation key identifies words starting with a lowercase letter in the author's name as an invalid part of an institute abbreviation. Think of the "and" in "National Aeronautics and Space Administration"). Since the author's name only consist of one "word", it is removed. (this is why the workaround @tobiasdiez mentioned does not work, even if it should)

#### Changes for
#6942
1. Identifying universities in the authors' names is arbitrary. The new regexp matches `^uni(v|b|$).*` instead of `^uni`. There are other suggestions in #6706 that can be implemented instead.
2. The inline institute abbreviation should now work again (e.g., `{JabRef ({JB})}` should produce `JB`)
3. I am ignoring the `shortauthor` field, because I find it unclear how and when to use the field (e.g., for multiple authors).

#7199 
1. I'll change how institutes are identified, so that the `Author` must only have a last name and that last name must contain a space. This should only happen if the `Author` originally were enclosed in brackets. In all other cases I'd argue it is acceptable to not abbreviate the author since the name is only one word.
    1. Drawbacks to this approach is that I am using the LatexToUnicodeAdapter to remove braces, and it doesn't like unbalanced ones.

#### Todo

- [x] Log a warning for unbalanced brackets in `Author`s
- [x] Change in CHANGELOG.md described (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

### CODE REVIEW

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Except for logging when LatexToUnicodeAdapter fails (otherwise names such as "{Link{\"{o}}ping University}**}**" will generate `Uni` and no warning, I don't intend any more code changes.

There will be some updates to the documentation, change log etc.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
Wait... I can't check for failed parsing?

https://github.com/tomtung/latex2unicode/blob/71979d6904fcf7efade786297983b881a1b019d4/src/main/scala/com/github/tomtung/latex2unicode/LaTeX2Unicode.scala#L139-L146

**Comment by tobiasdiez** [comment]
I guess you can use `parse` directly? Is this ready for review?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@tobiasdiez except for JavaDoc and changelog, yes, it is ready for review.

Which `parse`?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@tobiasdiez sorry, I missed to push a local commit.

**Comment by tobiasdiez** [comment]
https://github.com/tomtung/latex2unicode/blob/71979d6904fcf7efade786297983b881a1b019d4/src/main/scala/com/github/tomtung/latex2unicode/LaTeX2Unicode.scala#L126
this one?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
O.o
That would indeed work. Should I put this PR back in draft then? I am not sure how to go about this one, should I create a `LatexToUnicodeParseAdapter` that throws an exception so that the rest of the code in `LatexToUnicodeAdapter` is not repeated?

**Comment by tobiasdiez** [comment]
I have to admit I had not yet looked at your PR. It was just in response to your question. What do you want to do?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@tobiasdiez good then :)

In my opinion it is better if the generation warns if there is incorrect latex. But I am not sure how to do it in practice. I don't have experience with Scala <-> Java but if I am not misreading the `LaTeX2Unicode#parse` I am going to have to check if the result `instanceof fastparse.Success` in which case we need to add fastparse as a dependence?

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
I don't know the drawbacks/advantages of that approach, but if this is the only part of the JabRef code that needs this, perhaps it is better to leave the code as-is?

**Comment by tobiasdiez** [comment]
We already have farstparse as a dependency, so that's not a problem. I agree it would be nice to check for parsing errors, so I support you here (sadly only morally since I don't have any scala experience too).

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
fastparse is an indirect dependency from latex2unicode?
The issue I am having that `latex2unicode#parse` returns a `fastparse.core.Parsed` instance. Unless I am missing something it means I must be able to import `fastparse.core` which I cannot right now.
Should I `fastparse` to `module-info.java`?

**Comment by tobiasdiez** [comment]
yes, in principle importing it should work. However, there could be problems since fastparse is not yet modularized and we have a workaround in place: https://github.com/JabRef/jabref/blob/fd60d8394352f2cbfc9de2e70e3525d8bac1780e/build.gradle#L49-L50
So don't spend too much time trying to make it work.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@tobiasdiez you were right. For some reason my first attempt did not work.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
I need to check one more thing (`{European Union {\"{A}}viation Safety Agency}` -> `EUAeSA`) but I can't right now so I am putting this back in draft. Thank you for your help @tobiasdiez ! ❤️

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
I removed one test case because I believe is out of scope for this PR. How to deal with restricting the number of characters between the bracketedpattern/citationkey is a different issue (e.g., a bracketed pattern that should only have one character, Ö, can be accident turn into two characters, Oe, in the citation key generation).

File: src/main/java/org/jabref/logic/citationkeypattern/BracketedPattern.java [hunk 1]
```diff
@@ -1248,15 +1185,20 @@ private static String generateInstitutionKey(String content) {
...
-        result = unifyDiacritics(result);
-        result = result.replaceAll("^\\{", "").replaceAll("}$", "");
-        Matcher matcher = ABBREVIATIONS.matcher(result);
-        if (matcher.matches()) {
-            return matcher.group(1);
+        try {
+            result = LatexToUnicodeAdapter.format(content);
+        } catch (IllegalArgumentException e) {
+            LOGGER.warn(content + " could not be converted to unicode. This can result in an incorrect or missing institute citation key");
         }
```
Comment (k3KAW8Pnf7mkmdSMPHz27) [inline src/main/java/org/jabref/logic/citationkeypattern/BracketedPattern.java:1198]: This _should_ be caught earlier in the code but perhaps it is good to leave it in?

File: src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java [hunk 2]
```diff
@@ -5,23 +5,56 @@
...
+        try {
+            var parsingResult = LaTeX2Unicode.parse(toFormat);
+            if (parsingResult instanceof Parsed.Success) {
+                String text = parsingResult.get().value();
+                toFormat = Normalizer.normalize(text, Normalizer.Form.NFC);
+                return UNDERSCORE_PLACEHOLDER_MATCHER.matcher(toFormat).replaceAll("_");
+            } else {
+                throw new IllegalArgumentException("Parsing of latex failed.");
+            }
+        } catch (Throwable throwable) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:56]: I think it makes more sense to use  Optional<String> instead of exception handling for control flows (that is a bad idea).

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:56]: Essentially just replacing the `throw ...` with `return Optional.empty()`?

File: src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java [hunk 3]
```diff
@@ -5,23 +5,56 @@
...
+     */
+    public static String parse(String inField) throws IllegalArgumentException {
+        Objects.requireNonNull(inField);
+        String toFormat = UNDERSCORE_MATCHER.matcher(inField).replaceAll(REPLACEMENT_CHAR);
+        try {
+            var parsingResult = LaTeX2Unicode.parse(toFormat);
+            if (parsingResult instanceof Parsed.Success) {
+                String text = parsingResult.get().value();
+                toFormat = Normalizer.normalize(text, Normalizer.Form.NFC);
+                return UNDERSCORE_PLACEHOLDER_MATCHER.matcher(toFormat).replaceAll("_");
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:52]:
```suggestion
                return Optional.of(UNDERSCORE_PLACEHOLDER_MATCHER.matcher(toFormat).replaceAll("_"));
```

File: src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java [hunk 4]
```diff
@@ -5,23 +5,56 @@
...
+            } else {
+                throw new IllegalArgumentException("Parsing of latex failed.");
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:54]:
```suggestion
 return Optional.empty()
```

File: src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java [hunk 5]
```diff
@@ -5,23 +5,56 @@
...
+    }
+
+    /**
+     * Attempts to resolve all LaTeX in the String.
+     *
+     * @param inField a String containing LaTeX
+     * @return a String with LaTeX resolved into Unicode
+     * @throws IllegalArgumentException if the LaTeX could not be parsed
+     */
+    public static String parse(String inField) throws IllegalArgumentException {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:44]:
```suggestion
    public static Optional<String> parse(String inField) throws IllegalArgumentException {
```

File: src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java [hunk 6]
```diff
@@ -5,23 +5,56 @@
...
+     * @return a String with LaTeX resolved into Unicode, or the original String if the LaTeX could not be parsed
+     */
     public static String format(String inField) {
         Objects.requireNonNull(inField);
 
-        String toFormat = underscoreMatcher.matcher(inField).replaceAll(replacementChar);
-        toFormat = Normalizer.normalize(LaTeX2Unicode.convert(toFormat), Normalizer.Form.NFC);
-        return underscorePlaceholderMatcher.matcher(toFormat).replaceAll("_");
+        try {
+            return parse(inField);
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:31]:
```suggestion
            return parse(inField).orElse(Normalizer.normalize(inField, Normalizer.Form.NFC));
```
that should work

File: src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java (hunk 2 above)
Comment (Siedlerchr) [reply src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:56]: Yes, and Optional.of..(...) in the if block. I added some code suggestions

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:56]: Sure, I'll take a look at it. Thank you for the suggestion!
In theory, I like exceptions, but it does not make the as readable 😄

Comment (Siedlerchr) [reply src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:56]: Effective Java, Item 69: 
https://ahdak.github.io/blog/effective-java-part-9/
>Exceptions are, as their name implies, to be used only for exceptional conditions; they should never be used for ordinary control flow. A well designed API must not force its clients to use exceptions for ordinary control flow.

File: src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java [hunk 7]
```diff
@@ -5,23 +5,56 @@
...
+        try {
+            var parsingResult = LaTeX2Unicode.parse(toFormat);
+            if (parsingResult instanceof Parsed.Success) {
+                String text = parsingResult.get().value();
+                toFormat = Normalizer.normalize(text, Normalizer.Form.NFC);
+                return Optional.of(UNDERSCORE_PLACEHOLDER_MATCHER.matcher(toFormat).replaceAll("_"));
+            } else {
+ return Optional.empty()
+            }
+        } catch (Throwable throwable) {
```
Comment (Siedlerchr) [inline src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java:56]: the try and catch can now be removed

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
I’ll be looking at it shortly (dinner)

File: src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java (hunk 7 above)
Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java]: Wow, it literally took me 20 min to figure out what you meant 🤡 
Sure, I added the try-catch part based on https://github.com/tomtung/latex2unicode/blob/71979d6904fcf7efade786297983b881a1b019d4/src/main/scala/com/github/tomtung/latex2unicode/LaTeX2Unicode.scala#L143-L146
and assumed it needed to be there #scala-scared 😬

Comment (k3KAW8Pnf7mkmdSMPHz27) [reply src/main/java/org/jabref/model/strings/LatexToUnicodeAdapter.java]: So much prettier ✨ 😍

**Comment by Siedlerchr** [comment]
Is this ready for review? (I think it looks good)

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> Is this ready for review? (I think it looks good)

Yup, it is ready.

**Review by tobiasdiez — CHANGES_REQUESTED** [review]
The change look good to me. Thanks a lot!

Could you please also add the examples from the issues as test to verify that the issue doesn't reoccur. Thanks!

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@tobiasdiez good catch, apparently I did not examine my assumptions enough. My concerns regarding including it is,

1. Based on https://github.com/JabRef/jabref/pull/6706#issuecomment-682179624 and others,
> (i) when using in the bibtex key generator, the non-ascii characters are converted to a "good" representation (as in your example: Göd -> Goe)
> ...
> , because the bibtex tooling is not fully UTF8 (biblatex+biber is, but not the more commonly used bibtex tool).

my thoughts were that, to the largest extent possible, BibTeX citation keys should be ASCII/ANSI only, in which case #7199 should not be added as a test case. If `{eBay}` is successful, it is likely that #7199 will be as well (from BibTeX/Java´s point-of-view both contain only a last name that starts with a lowercase letter, and it was the institute abbreviation method that removed 杨秀群).

2. I am not convinced that the currently generated citation key is correct, but you seem to know more about this particular issue than I, so feel free to correct me 😛 
I based my decision on @xiaodongcentury's comment 
> "杨 秀群", but not "杨秀群", the former has a space between last name and first name of a Chinese name.

In my opinion, the "correct" citation key should be either 杨 or 秀群 (i.e., based on the documentation, the citation key shoul be the last name). With `eBay` I know the citation key is correct.

However, all of that being said (and my apologies about the rather lengthy reading), if this is how JabRef is used, perhaps the test case should be added anyway?

**Comment by tobiasdiez** [comment]
As we have quite a lot of Chinese users, it would be good to support this as far as possible. Thus, if the user enters Chinese authors, then I would say we should generate a key with chinese characters. In the best case, one would actually replace the Chinese character by their romanization, but I'm not sure if there exist a convenient library for this.

For now, I would say we simply add `杨秀群` as a test (with whatever result it currently produces).

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
> For now, I would say we simply add `杨秀群` as a test (with whatever result it currently produces).

Sure!

> In the best case, one would actually replace the Chinese character by their romanization, but I'm not sure if there exist a convenient library for this.

[pinyin4j](http://pinyin4j.sourceforge.net/) seem to be the most popular Java alternative. However, given that there isn't a need of high performance (clean up operation?), [there are more popular javascript libraries](https://github.com/hotoo/pinyin/blob/master/README-us_EN.md). I don't know if there is an interest for this?

**Comment by tobiasdiez** [comment]
Thanks, also for the research concerning pinyin libraries. Let's see if users report this as a feature request.

**Comment by k3KAW8Pnf7mkmdSMPHz27** [comment]
@Siedlerchr @tobiasdiez thank you both for your time, reviews and suggestions, they are appreciated!

**Comment by xiaodongcentury** [comment]
Thank you ALL!
The JabRef 5.3--2020-12-28--78b08b5 works perfectly with Chinese names.
Thanks again.
